### PR TITLE
bugfix correct error backoff handling for lb gateway

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <kpango@vdaas.org>

### Description:
I fixed correct error backoff handling for lb gateway

1. old status is below
LBGW never backoff on current implementation in all case,  it was not working correctly.
2. the bug detail is below.
In the gRPC Handler of LBGW, when BroadCast or DoMulti is executed, if it does not return an Error, the backoff of internal/net/grpc/client will not work, but LBGW does not return an Error using return nil.
3. how to fixed.
LBGW now returns the correct error to internal/net/grpc/client, but this client also backs off in some cases, such as ContextCancel. To address this issue, we have improved gRPC Status and fixed the backoff condition.

### Environment:

- Go Version: 1.16.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [x] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by vankichi — DISMISSED** [review]

**Comment by rinx** [comment]
What was the problem? Please describe the bug.

**Comment by kpango** [comment]
@rinx the problem is LBGateway never backoff on current implementation.
so I fixed `correct error backoff handling for lb gateway`

**Comment by rinx** [comment]
I already know about that.
I don't understand what case doesn't it work correctly.

**Comment by kpango** [comment]
In all cases, it was not working correctly.

**Comment by kpango** [comment]
@rinx FYI
In the gRPC Handler of LBGW, when BroadCast or DoMulti is executed, if it does not return an Error, the backoff of internal/net/grpc/client will not work, but LBGW does not return an Error using return nil.

**Comment by kpango** [comment]
I updated PR description please check it.

**Comment by rinx** [comment]
Thanks. I finally understand the point to review.

File: pkg/gateway/lb/handler/grpc/handler.go [hunk 1]
```diff
@@ -105,7 +109,7 @@ func (s *server) Exists(ctx context.Context, meta *payload.Object_ID) (id *paylo
 				if sspan != nil {
 					sspan.SetStatus(trace.FromGRPCStatus(st.Code(), msg))
 				}
-				return nil
+				return err
```
Comment (rinx) [inline pkg/gateway/lb/handler/grpc/handler.go:112]: I'm not sure but I think it should not return an error if it is a NotFound error.

File: pkg/gateway/lb/handler/grpc/handler.go [hunk 2]
```diff
@@ -888,7 +893,7 @@ func (s *server) Insert(ctx context.Context, req *payload.Insert_Request) (ce *p
 				}
 				emu.Unlock()
 			}
-			return nil
+			return err
```
Comment (rinx) [inline pkg/gateway/lb/handler/grpc/handler.go:896]: I think it should not return an error if it is an AlreadyExists error.

File: pkg/gateway/lb/handler/grpc/handler.go [hunk 3]
```diff
@@ -2098,7 +2103,7 @@ func (s *server) GetObject(ctx context.Context, req *payload.Object_VectorReques
 				if span != nil {
 					span.SetStatus(trace.FromGRPCStatus(st.Code(), msg))
 				}
-				return nil
+				return err
```
Comment (rinx) [inline pkg/gateway/lb/handler/grpc/handler.go:2106]: Same. A NotFound error is ignorable.

**Review by rinx — COMMENTED** [review]
Thanks for fixing the bug.
I added several comments but I'm not sure they are correct.

**Comment by kpango** [comment]
@rinx thank you for reviewing.
I fixed the point that you pointed out.

FYI: vald's internal grpc client only backoffs when status is Internal, Resource-Exhausted, Unavailable.
https://github.com/vdaas/vald/blob/master/internal/net/grpc/client.go#L606-L611

**Review by vankichi — APPROVED** [review]

**Review by rinx — APPROVED** [review]

**Pull request merged**